consensus/hotstuff/timeoutcollector/timeout_cache.go

```diff
@@ -0,0 +1,108 @@
+	lock     sync.RWMutex
+	view     uint64
+	timeouts map[flow.Identifier]*model.TimeoutObject // signerID -> first timeout
+}
+
+// NewTimeoutObjectsCache instantiates a TimeoutObjectsCache for the given view
+func NewTimeoutObjectsCache(view uint64) *TimeoutObjectsCache {
+	return &TimeoutObjectsCache{
+		view:     view,
+		timeouts: make(map[flow.Identifier]*model.TimeoutObject),
+	}
+}
+
+func (vc *TimeoutObjectsCache) View() uint64 { return vc.view }
+
+// AddTimeoutObject stores a timeout in the cache. The following errors are expected during
+// normal operations:
+//  * nil: if the vote was successfully added
+//  * model.DoubleTimeoutError is returned if the voter is equivocating
+//  * RepeatedTimeoutErr is returned when adding a timeout for the same view from
+//    the same voter multiple times.
+//  * TimeoutForIncompatibleViewError is returned if the timeout is for a different view.
+// When AddTimeoutObject returns an error, the timeout is _not_ stored.
+func (vc *TimeoutObjectsCache) AddTimeoutObject(timeout *model.TimeoutObject) error {
+	if timeout.View != vc.view {
+		return TimeoutForIncompatibleViewError
+	}
+	vc.lock.Lock()
+	defer vc.lock.Unlock()
+
+	// De-duplicated timeouts based on the following rules:
+	//  * Vor each voter (i.e. SignerID), we store the _first_  t0.
+	//  * For any subsequent timeout t, we check whether t.ID() == t0.ID().
+	//    If this is the case, we consider the timeout a duplicate and drop it.
+	//    If t and t0 have different checksums, the voter is equivocating, and
+	//    we return a model.DoubleTimeoutError.
+	firstTimeout, exists := vc.timeouts[timeout.SignerID]
+	if exists {
+		if firstTimeout.ID() != timeout.ID() {
+			return model.NewDoubleTimeoutErrorf(firstTimeout, timeout, "detected timeout equivocation at view: %d", vc.view)
+		}
+		return RepeatedTimeoutErr
+	}
+	vc.timeouts[timeout.SignerID] = timeout
+
+	return nil
+}
+
+// GetTimeoutObject returns the stored timeout for the given `signerID`. Returns:
+//  - (timeout, true) if a timeout object from signerID is known
+//  - (false, nil) no timeout object from signerID is known
+func (vc *TimeoutObjectsCache) GetTimeoutObject(signerID flow.Identifier) (*model.TimeoutObject, bool) {
+	vc.lock.RLock()
+	timeout, exists := vc.timeouts[signerID] // if signerID is unknown, its `Vote` pointer is nil
+	vc.lock.RUnlock()
+	return timeout, exists
+}
+
+// Size returns the number of cached timeout objects
+func (vc *TimeoutObjectsCache) Size() int {
+	vc.lock.RLock()
+	s := len(vc.timeouts)
+	vc.lock.RUnlock()
+	return s
+}
+
+// All returns all currently cached timeout objects. Concurrency safe.
+func (vc *TimeoutObjectsCache) All() []*model.TimeoutObject {
+	vc.lock.Lock()
+	defer vc.lock.Unlock()
+	return vc.all()
+}
+
+// all returns all currently cached timeout objects. NOT concurrency safe
+func (vc *TimeoutObjectsCache) all() []*model.TimeoutObject {
+	timeoutObjects := make([]*model.TimeoutObject, len(vc.timeouts))
```

Comment:
```suggestion
	timeoutObjects := make([]*model.TimeoutObject, 0, len(vc.timeouts))
```

Else `timoutObjects` will be `[nil,nil,nil,...,TO1,TO2,TO3,...]`

Reply:
good catch